### DISCUSSION TITLE
add_driver_ahead gives ValueError due to length mismatch with position data

### DISCUSSION THREAD

**Discussion by man-shar** [theOehrly/Fast-F1#172] [Q&A]
Hi folks,

New user of this awesome API here.

I'm trying to add the driver ahead column like so:
`lap2 = driver_laps.iloc[1]`
`lap2_data = driver_data.slice_by_lap(lap2)`
`lap2_data.add_driver_ahead()`

But keep getting an error saying:
`ValueError: Length of values (362) does not match length of index (369)`

I checked the lengths of the `slice_by_lap()` result and the `calculate_driver_ahead()` array, and they do have a difference in lengths as shown in the error (362 vs 369).

Wondering if anyone has faced this issue before and how you solved it?

Thanks,
Manas

**Comment by theOehrly** [comment]
Which session, lap, ... does produce that error?

**Comment by man-shar** [comment]
Hi, 

All of them, really. You can try Verstappen-Lap 1-2021 Bahrain. Or Hamilton-Lap 1-2021 Bahrain. 

I _did_ manage to add driver ahead by running the legacy function `fastf1.legacy.inject_driver_ahead`, which had no errors but is slow (is the new one faster?).

Wondering if it's something to do with having integrated distance/position values?

Thanks!

**Reply by theOehrly** [reply]
Can you give a short code example please? I am failing to reproduce this. Furthermore, plenty of people use the `.get_telemetry` function, which internally calls `.add_driver_ahead`. You say this problem would occur with many laps. I'm not sure how this could have gone unnoticed for so long though. 

The legacy implementation is slower, yes. But the new implementation also produces a smoother result. See the comparison plot here: https://theoehrly.github.io/Fast-F1/legacy.html

**Reply by man-shar** [reply]
Thanks!
And yeah. Think you're right. I might be doing it in a convoluted way. Here's what I'm doing:


```python
import fastf1

fastf1.Cache.enable_cache('./cached-race-data')

races_2021 = fastf1.get_event_schedule(2021)

for i, race_metadata in races_2021.iterrows():
    race_data = fastf1.get_session(
        race_metadata["EventDate"].year, race_metadata["EventName"], "R")

    race_data.load(telemetry=True, laps=True)

    pos_data = race_data.pos_data
    laps = race_data.laps

    for driver_number, driver_data in pos_data.items():
      # just try verstappen for now
        if(driver_number != "33"):
            continue

        driver_laps = laps[laps["DriverNumber"]
                           == driver_number].reset_index()

        for j, lap in driver_laps.iterrows():
            lap_number = lap["LapNumber"]

            lap_data = driver_data.slice_by_lap(
                driver_laps.iloc[j])

            # >>> ERROR here: Length of values (443) does not match length of index (486)
            lap_data.add_driver_ahead()

```

**Reply by man-shar** [reply]
okay so I tried the way it is on the link you shared and it works perfectly fine. thanks a ton!!

It seems like lap.get_car_data() and driver_data.slice_by_lap(lap) return different data frames?

**Reply by theOehrly** [reply]
Seeing your code now, the problem is fairly obvious. You are calling ´.add_driver_ahead´ on the pos_data. But the distance to driver ahead is calculated from car_data (Speed, Distance). Therefore, there is a mismatch, because car_data and pos_data do not use the same time base.

This is a bug actually, I'd say. But usually no one runs into this problem, because the way this is called internally, the problem does not occur.

**Reply by man-shar** [reply]
Ah I see. Thanks a bunch for helping out!

Is there interest/value in fixing that bug/inconsistency with the time base? Admittedly, I'm not well versed with library internals but won't mind digging in and sending a PR?

**Reply by theOehrly** [reply]
> Ah I see. Thanks a bunch for helping out!
> 
> Is there interest/value in fixing that bug/inconsistency with the time base? Admittedly, I'm not well versed with library internals but won't mind digging in and sending a PR?

Sure, PRs are always welcome. I opened an issue (#178) for this with a shortened example and I also proposed an idea for how to fix it there. Feel free to take a look at it. And if you have questions, just ask them in that issue.

**Reply by theOehrly** [reply]
Just FYI, this will be fixed in the next release, in case it is relevant for you.

**Reply by man-shar** [reply]
thank you!
